WIP ARROW-4787: [C++] Add support for Null in MemoTable and related kernels

fsaintjacques opened the pull request.

pitrou commented:
@fsaintjacques Can you fix the Python tests?

pitrou commented on cpp/src/arrow/util/hashing.h:
```diff
@@ -698,6 +808,23 @@ struct HashTraits<T, enable_if_fixed_size_binary<T>> {
+    *null_count = 1;
+    RETURN_NOT_OK(
+        internal::BitmapAllButOne(pool, dict_length, null_index, validity_bitmap));
+  }
```
Same for `*null_count = 0`.

fsaintjacques replied:
It is done by the caller, but I will also explicit it in the function (compiler should optimize it away).

pitrou commented on cpp/src/arrow/util/hashing.h:
```diff
@@ -698,6 +808,23 @@ struct HashTraits<T, enable_if_fixed_size_binary<T>> {
 };
 
+template <typename MemoTableType>
+static inline Status ComputeValidityBitmap(MemoryPool* pool,
```
We usually call it "null bitmap" ;-)

wesm replied:
"validity bitmap" is used in some other places (e.g. in the Arrow spec itself), it's my fault for using the "null bitmap" terminology in the first place in this codebase

fsaintjacques commented on cpp/src/arrow/util/hashing.h:
```diff
@@ -35,6 +34,7 @@
 #include "arrow/array.h"
 #include "arrow/buffer.h"
 #include "arrow/builder.h"
+#include "arrow/compute/kernels/util-internal.h"
```
This is a leftover from a refactor I did. I'll remove it.

fsaintjacques commented on cpp/src/arrow/compute/kernels/hash-test.cc:
```diff
@@ -160,8 +161,8 @@ TYPED_TEST(TestHashKernelPrimitive, DictEncode) {
-                                {true, false, true, true, true, true}, {2, 1, 3}, {},
-                                {0, 0, 0, 1, 0, 2});
+                                {true, false, true, true, true, true}, {2, 0, 1, 3},
+                                {1, 0, 1, 1}, {0, 0, 0, 2, 0, 3});
```
Once this is merged, I need to open a discussion on the mailing list as the format spec doesn't mandate any of this.

pitrou replied:
Right. But for the time being, we should probably choose one or the other. I don't think it makes sense to add a null to the dict values if it's not referred to by the indices.
(I think for compatibility we should perhaps keep the current behaviour @wesm )

fsaintjacques replied:
I'm working on keeping the current behavior (which fill fix the python test).

wesm replied:
I don't think this is a specification-related matter. The data is simply data

fsaintjacques commented:
@pitrou I reverted the DictEncode behaviour, thus no need to change the python tests (and also pandas categorical conversion). While doing so, I refactored the HashKernel a bit since there's 2 bit provided at compile time. 

I'm still not satisfied, but it'll do for now. I believe we'll have to break the HashKernel/Action implementation into specific kernels soon.

pitrou commented:
It looks like the Python tests still need a fix (see AppVeyor results).
